### PR TITLE
Migrate contacts

### PR DESCRIPTION
**Pull request by DeepDiver1975**
- [x] requires https://github.com/owncloud/core/pull/21540
# Test Scenario:
1. Export the database tables(oc_contacts_addressbooks and oc_contacts_cards) from an existing instance which has the contacts app in use
2. import these tables into the database you use for testing
3. run the migration

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
... will continue next year ... :zzz:

**Comment by DeepDiver1975** [comment]
Before I move on here we need to decide on having a fully automated migration or give users and admins the opportunity to migrate their data themselves.

@karlitschek @georgehrke @jancborchardt @jospoortvliet

**Comment by georgehrke** [comment]
Well, the old calendar and contacts app won't work with ownCloud 9 anymore, so if one used these apps, which is the case if there is data in the corresponding db tables, the data should automatically be migrated to the new solution, so one can keep using ownCloud for caldav and carddav. No need to know which app stores contacts and calendar.

**Comment by DeepDiver1975** [comment]
Well - I basically see two paths we can walk:
1. kill the old apps, automatically migrate data and provide new client web apps via the appstore on release date
2. ensure the olds apps continue to work and allow admins and users to manually migrate and decide when to move over

Both scenarios have their pros and cons - but the first scenario enforces a hard move of all instances in this world. 
Are we feature complete with the new apps compared to the new apps?
Are we even able to finish calendar and contacts rework in time?
We have no fallback in case there are undiscovered issues - the old apps are broken.

Furthermore: there is one mandatory manual migration step: every client has to be setup again from scratch! I see no option to automatically migrate client setups to use the new webdav url.

**Comment by karlitschek** [comment]
I agree with @DeepDiver1975 
Ideally we go with option 1. This is the ultimate goal.
But if we come to the conclusion that the new Contacts and Calendar are not production ready yet for 9.0 then we have to go with option 2 which unfortunately makes a lot of things more complicated.

**Comment by DeepDiver1975** [comment]
Okay - so - let's aim for a occ based approach first which we can later extend to an fully automated or ui based approach.

**Comment by jancborchardt** [comment]
> Furthermore: there is one mandatory manual migration step: every client has to be setup again from scratch! I see no option to automatically migrate client setups to use the new webdav url.

Ah damn, right … can we not mock / redirect these URLs?

**Comment by DeepDiver1975** [comment]
> Ah damn, right … can we not mock / redirect these URLs?

no - this completely violates the idea of URIs ... but even if - we cannot test all clients out there if they properly can handle these redirects

**Comment by karlitschek** [comment]
instead of redirects, can we maybe just register routes for the old urls too? So that the old once work too and users can migrate over over time? Same way we did for webdav a few years ago?

**Comment by DeepDiver1975** [comment]
> instead of redirects, can we maybe just register routes for the old urls too? So that the old once work too and users can migrate over over time? Same way we did for webdav a few years ago?

I see this really risky because we don't migrate the logic as well - e.g. we have no birthday calendar for contacts, the url logic for shared addressbooks and calendars is different and what so ever ...

... and again - I doubt that the clients can properly deal with this change

**Comment by karlitschek** [comment]
Hmm. This is a real problem for migration. We basically the address book syncing will stop for all users when they upgrade to 9.0 and they have to manually figure out what is going on and what to do.
Let's talk on Thursday what the best option is.

**Comment by PVince81** [comment]
So I guess you guys discussed the best option and from the commits it looks like there's an app that does the migration

**Comment by PVince81** [comment]
How to test ?

**Comment by DeepDiver1975** [comment]
> So I guess you guys discussed the best option and from the commits it looks like there's an app that does the migration

the initial idea was to do it as an app - now this is all part of the dav app.

**Comment by DeepDiver1975** [comment]
> How to test ?

I just updated the description

**Comment by PVince81** [comment]
@DeepDiver1975 you forgot one important step: how to verify it worked :wink:

**Comment by DeepDiver1975** [comment]
> you forgot one important step: how to verify it worked :wink:

happy to listen to suggestions

**Comment by PVince81** [comment]
I don't know... didn't find how to verify that's why I ask.

The code seems to provide a new CLI command, so do I need to run it for the migration or is it automatic ? For some reason I didn't find it in the "occ" list of commands.
I didn't dig deep enough in the code to find this out.

**Comment by DeepDiver1975** [comment]
> The code seems to provide a new CLI command, so do I need to run it for the migration or is it automatic

the migration has to/will run on upgrade from 8.2 to 9.0 - for testing I provided an occ command which is only available in debug mode.

sorry - I should have written this down here ..... maybe I did add this somewhere else .....

**Comment by rullzer** [comment]
Tested with some manually added contacts. And seems to work.
I'll try to test with a bit set of cards later.

File: apps/dav/command/migrateaddressbooks.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+			->setDescription('Migrate addressbooks from the contacts app to core')
+			->addArgument('user',
+				InputArgument::OPTIONAL,
+				'User for whom all addressbooks will be migrated');
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$this->service->setup();
+
+		if ($input->hasArgument('user')) {
```
Comment (MorrisJobke) [inline apps/dav/command/migrateaddressbooks.php]: Doesn't seem to trigger:

```
$ ./occ migrate:addressbooks

                              
  [InvalidArgumentException]  
  User <> in unknown.         
                              

migrate:addressbooks [<user>]
```

**Comment by MorrisJobke** [comment]
I would debug further, but it doesn't migrate anymore after this exception happened:

```
$ ./occ migrate:addressbooks mjob                                                                                                                                                                           1 ↵

                                                                                                                                                                                                                                                                               
  [Doctrine\DBAL\Exception\SyntaxErrorException]                                                                                                                                                                                                                               
  An exception occurred while executing 'SELECT  FROM `oc_share` WHERE (`item_source` = ?) AND (`item_type` = 'addressbook') AND (`share_type` IN ('0', '1'))' with params ["16"]:                                                                                             
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM `oc_share` WHERE (`item_source` = '16') AND (`item_type` = 'addressbook  
  ') A' at line 1                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                               

                                                                                                                                                                                                                                                                               
  [Doctrine\DBAL\Driver\PDOException]                                                                                                                                                                                                                                          
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM `oc_share` WHERE (`item_source` = '16') AND (`item_type` = 'addressbook  
  ') A' at line 1                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                               

                                                                                                                                                                                                                                                                               
  [PDOException]                                                                                                                                                                                                                                                               
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM `oc_share` WHERE (`item_source` = '16') AND (`item_type` = 'addressbook  
  ') A' at line 1                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                               

migrate:addressbooks [<user>]

```

**Comment by MorrisJobke** [comment]
I dropped everything form oc_addressbooks, oc_addressbookchanges, oc_cards and oc_cards_properties and now have the full stack traces:

```
$ ./occ migrate:addressbooks mjob -vvv

                                                                                                                                                                                                                                                                               
  [Doctrine\DBAL\Exception\SyntaxErrorException]                                                                                                                                                                                                                               
  An exception occurred while executing 'SELECT  FROM `oc_share` WHERE (`item_source` = ?) AND (`item_type` = 'addressbook') AND (`share_type` IN ('0', '1'))' with params ["16"]:                                                                                             
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM `oc_share` WHERE (`item_source` = '16') AND (`item_type` = 'addressbook  
  ') A' at line 1                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                               

Exception trace:
 () at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:90
 Doctrine\DBAL\Driver\AbstractMySQLDriver->convertException() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php:116
 Doctrine\DBAL\DBALException::driverExceptionDuringQuery() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:836
 Doctrine\DBAL\Connection->executeQuery() at /Users/morrisjobke/Projects/owncloud/master/lib/private/db/connection.php:182
 OC\DB\Connection->executeQuery() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:206
 Doctrine\DBAL\Query\QueryBuilder->execute() at /Users/morrisjobke/Projects/owncloud/master/lib/private/db/querybuilder/querybuilder.php:124
 OC\DB\QueryBuilder\QueryBuilder->execute() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/addressbookadapter.php:82
 OCA\Dav\Migration\AddressBookAdapter->getShares() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:73
 OCA\Dav\Migration\MigrateAddressbooks->migrateShares() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:50
 OCA\Dav\Migration\MigrateAddressbooks->OCA\Dav\Migration\{closure}() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/addressbookadapter.php:47
 OCA\Dav\Migration\AddressBookAdapter->foreachBook() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:51
 OCA\Dav\Migration\MigrateAddressbooks->migrateForUser() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/command/migrateaddressbooks.php:49
 OCA\Dav\Command\MigrateAddressbooks->execute() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:840
 Symfony\Component\Console\Application->doRunCommand() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:192
 Symfony\Component\Console\Application->doRun() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:123
 Symfony\Component\Console\Application->run() at /Users/morrisjobke/Projects/owncloud/master/lib/private/console/application.php:110
 OC\Console\Application->run() at /Users/morrisjobke/Projects/owncloud/master/console.php:81
 require_once() at /Users/morrisjobke/Projects/owncloud/master/occ:11

                                                                                                                                                                                                                                                                               
  [Doctrine\DBAL\Driver\PDOException]                                                                                                                                                                                                                                          
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM `oc_share` WHERE (`item_source` = '16') AND (`item_type` = 'addressbook  
  ') A' at line 1                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                               

Exception trace:
 () at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:93
 Doctrine\DBAL\Driver\PDOStatement->execute() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:828
 Doctrine\DBAL\Connection->executeQuery() at /Users/morrisjobke/Projects/owncloud/master/lib/private/db/connection.php:182
 OC\DB\Connection->executeQuery() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:206
 Doctrine\DBAL\Query\QueryBuilder->execute() at /Users/morrisjobke/Projects/owncloud/master/lib/private/db/querybuilder/querybuilder.php:124
 OC\DB\QueryBuilder\QueryBuilder->execute() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/addressbookadapter.php:82
 OCA\Dav\Migration\AddressBookAdapter->getShares() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:73
 OCA\Dav\Migration\MigrateAddressbooks->migrateShares() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:50
 OCA\Dav\Migration\MigrateAddressbooks->OCA\Dav\Migration\{closure}() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/addressbookadapter.php:47
 OCA\Dav\Migration\AddressBookAdapter->foreachBook() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:51
 OCA\Dav\Migration\MigrateAddressbooks->migrateForUser() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/command/migrateaddressbooks.php:49
 OCA\Dav\Command\MigrateAddressbooks->execute() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:840
 Symfony\Component\Console\Application->doRunCommand() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:192
 Symfony\Component\Console\Application->doRun() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:123
 Symfony\Component\Console\Application->run() at /Users/morrisjobke/Projects/owncloud/master/lib/private/console/application.php:110
 OC\Console\Application->run() at /Users/morrisjobke/Projects/owncloud/master/console.php:81
 require_once() at /Users/morrisjobke/Projects/owncloud/master/occ:11

                                                                                                                                                                                                                                                                               
  [PDOException]                                                                                                                                                                                                                                                               
  SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FROM `oc_share` WHERE (`item_source` = '16') AND (`item_type` = 'addressbook  
  ') A' at line 1                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                               

Exception trace:
 () at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91
 PDOStatement->execute() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91
 Doctrine\DBAL\Driver\PDOStatement->execute() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Connection.php:828
 Doctrine\DBAL\Connection->executeQuery() at /Users/morrisjobke/Projects/owncloud/master/lib/private/db/connection.php:182
 OC\DB\Connection->executeQuery() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php:206
 Doctrine\DBAL\Query\QueryBuilder->execute() at /Users/morrisjobke/Projects/owncloud/master/lib/private/db/querybuilder/querybuilder.php:124
 OC\DB\QueryBuilder\QueryBuilder->execute() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/addressbookadapter.php:82
 OCA\Dav\Migration\AddressBookAdapter->getShares() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:73
 OCA\Dav\Migration\MigrateAddressbooks->migrateShares() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:50
 OCA\Dav\Migration\MigrateAddressbooks->OCA\Dav\Migration\{closure}() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/addressbookadapter.php:47
 OCA\Dav\Migration\AddressBookAdapter->foreachBook() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/lib/migration/migrateaddressbooks.php:51
 OCA\Dav\Migration\MigrateAddressbooks->migrateForUser() at /Users/morrisjobke/Projects/owncloud/master/apps/dav/command/migrateaddressbooks.php:49
 OCA\Dav\Command\MigrateAddressbooks->execute() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:840
 Symfony\Component\Console\Application->doRunCommand() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:192
 Symfony\Component\Console\Application->doRun() at /Users/morrisjobke/Projects/owncloud/master/3rdparty/symfony/console/Application.php:123
 Symfony\Component\Console\Application->run() at /Users/morrisjobke/Projects/owncloud/master/lib/private/console/application.php:110
 OC\Console\Application->run() at /Users/morrisjobke/Projects/owncloud/master/console.php:81
 require_once() at /Users/morrisjobke/Projects/owncloud/master/occ:11

migrate:addressbooks [<user>]
```

**Comment by DeepDiver1975** [comment]
note to self: unit test getShares

**Comment by DeepDiver1975** [comment]
@MorrisJobke please have a look - thx

**Comment by MorrisJobke** [comment]
Tested and works with my production data \o/ :+1:

File: apps/dav/command/migrateaddressbooks.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
+<?php
+
```
Comment (LukasReschke) [inline apps/dav/command/migrateaddressbooks.php]: Missing copyright header

Comment (LukasReschke) [inline apps/dav/command/migrateaddressbooks.php]: Added with https://github.com/owncloud/core/commit/128d9d7aa624cc013cb8f0dbb79693ee5d6c68dd

**Comment by LukasReschke** [comment]
:+1:

**Pull request merged**